crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -83,47 +121,55 @@ impl Plugin for VisibilityPlugin {
-                .after(TransformSystem::TransformPropagate),
-        )
-        .add_system_to_stage(
-            CoreStage::PostUpdate,
-            update_frusta::<Projection>
-                .label(UpdateProjectionFrusta)
-                .after(TransformSystem::TransformPropagate),
-        )
-        .add_system_to_stage(
-            CoreStage::PostUpdate,
-            check_visibility
-                .label(CheckVisibility)
-                .after(CalculateBounds)
-                .after(UpdateOrthographicFrusta)
-                .after(UpdatePerspectiveFrusta)
-                .after(UpdateProjectionFrusta)
-                .after(TransformSystem::TransformPropagate),
-        );
+        app.init_resource::<EntityMeshRelationships>()
+            .add_system_to_stage(
+                CoreStage::PostUpdate,
+                calculate_bounds.label(CalculateBounds),
+            )
+            .add_system_to_stage(CoreStage::PostUpdate, update_bounds.label(UpdateBounds))
+            .add_system_to_stage(
+                CoreStage::PostUpdate,
+                update_frusta::<OrthographicProjection>
+                    .label(UpdateOrthographicFrusta)
+                    .after(TransformSystem::TransformPropagate),
+            )
+            .add_system_to_stage(
+                CoreStage::PostUpdate,
+                update_frusta::<PerspectiveProjection>
+                    .label(UpdatePerspectiveFrusta)
+                    .after(TransformSystem::TransformPropagate),
+            )
+            .add_system_to_stage(
+                CoreStage::PostUpdate,
+                update_frusta::<Projection>
+                    .label(UpdateProjectionFrusta)
+                    .after(TransformSystem::TransformPropagate),
+            )
+            .add_system_to_stage(
+                CoreStage::PostUpdate,
+                check_visibility
+                    .label(CheckVisibility)
+                    .after(CalculateBounds)
+                    .after(UpdateOrthographicFrusta)
+                    .after(UpdatePerspectiveFrusta)
+                    .after(UpdateProjectionFrusta)
+                    .after(TransformSystem::TransformPropagate),
+            );
     }
 }
 
+/// Calculates [`Aabb`]s for [`Entities`](Entity) with [`Mesh`]es. To opt out of bound calculation
+/// for an `Entity`, give it the [`NoFrustumCulling`] component.
 pub fn calculate_bounds(
     mut commands: Commands,
     meshes: Res<Assets<Mesh>>,
     without_aabb: Query<(Entity, &Handle<Mesh>), (Without<Aabb>, Without<NoFrustumCulling>)>,
+    mut entity_mesh_rel: ResMut<EntityMeshRelationships>,
 ) {
     for (entity, mesh_handle) in without_aabb.iter() {
+        // Record entities whose Aabb's we're managing.
+        entity_mesh_rel.register(entity, mesh_handle);
+
         if let Some(mesh) = meshes.get(mesh_handle) {
             if let Some(aabb) = mesh.compute_aabb() {
                 commands.entity(entity).insert(aabb);
```

Comment:
```suggestion
        if let Some(mesh) = meshes.get(mesh_handle) {
            if let Some(aabb) = mesh.compute_aabb() {
                entity_mesh_rel.register(entity, mesh_handle);
                commands.entity(entity).insert(aabb);
```
the call to `register` should probably be added here, so that it mirrors the actual state of the ECs if one of the `if let` fails.

Reply:
Oh actually ... hmmm ... I'm not so sure. I have two competing thoughts.

**Why It Should Be Inside the If Lets**
If an entity has a mesh and does not have an Aabb then it'll be in this query. If the assets collection doesn't have a mesh for the entity's handle or the Aabb can't be computed then this entity will be here the next time this system is called. We don't want to add it to the relationship map twice.

**Why It Should Be Outside the If Lets**
`entity_mesh_rel` tracks which entities have which meshes. If we fail to compute an Aabb for a mesh that doesn't mean that the entity _doesn't_ have that mesh. So we should record it regardless

That being said I'm leaning towards doing what you suggest but it might help if I knew more about the failure states of `meshes.get` and `compute_aabb` :thinking:

Reply:
Actually yeah I guess it's good to not add things until we compute an `Aabb` for it. If an entity has a mesh but not an Aabb and then its mesh is mutated we probably don't want to try to assign a mesh _and_ try and update the mesh :thinking:

Reply:
Updated in b02be5c18b8f29a6ac8aac6bda3dc466dffed11a.

Reply:
`meshes.get()` is likely to fail at least once. `Handle<Mesh>`s are added to entities before the mesh is fully loaded, and it's possible that a few frames before the mesh gets loaded that it returns `None`.

`compute_aabb` should only fail if the mesh doesn't have `ATTRIBUTE_POSITION` or has malformed vertex positions. bevy supports custom mesh attributes, so I think it's possible some meshes will have their `compute_aabb` fail all the time. The more I think about this, the more I believe the `aabb` should be part of the `Mesh`. But that's for another time.